alf/algorithms/hypernetwork_algorithm.py

```diff
@@ -0,0 +1,479 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl import logging
+import gin
+import numpy as np
+import torch
+import torch.nn.functional as F
+from typing import Callable
+
+import alf
+from alf.data_structures import AlgStep, LossInfo, namedtuple
+from alf.algorithms.generator import Generator
+from alf.algorithms.hypernetwork_networks import ParamNetwork
+from alf.networks import EncodingNetwork
+from alf.tensor_specs import TensorSpec
+from alf.utils import common, math_ops
+from alf.utils.summary_utils import record_time
+
+HyperNetworkLossInfo = namedtuple("HyperNetworkLossInfo", ["loss", "extra"])
+
+
+@gin.configurable
+class HyperNetwork(Generator):
```

Comment:
I feel HyperNetwork should be a subclass of ``Generator``. The output of ``Generator`` is the result directly from the generator. While the results of ``HyperNetwork`` is the results of the generated network applied on inputs.
I feel it might be better if the HyperNetwork is implemented by using a generator and applying proper likelihood function (which is neural network with some loss in this case).